### PR TITLE
user defined blacklist

### PR DESCRIPTION
**Pull request by sergey-safarov**
Added user defined blacklist.
Blacklist may used as white list.
Blacklist applied to callflow module/actions. Also added blacklist cache.
Defined `skip_human` and `ask_human` new blacklist actions.

Feature wants to resolve FS-11894
https://freeswitch.org/jira/browse/FS-11894

### CODE REVIEW

**Comment by sergey-safarov** [comment]
Warning: github list commit in wrong order.
If you want look in correct order, then please use this [link](https://github.com/sergey-safarov/kazoo/commits/user_blacklist3)

**Comment by sergey-safarov** [comment]
I see `kazoo_endpoints` added into FreeeSwitch mod_kazoo module.
I checked `group_confirm_key` feature works properly when call originated like
```
[group_confirm_key='1',group_confirm_cancel_timeout='true',group_confirm_read_timeout='7000',group_confirm_file='ivr/ivr-welcome_to.wav',ignore_early_media='true']kz/sip:user@domain 
```

File: applications/callflow/src/callflow.hrl [hunk 1]
```diff
@@ -6,6 +6,13 @@
 -include_lib("kazoo_call/include/kapps_call_command_types.hrl").
 -include_lib("kazoo_documents/include/kazoo_documents.hrl").
 
+-record(cf_bl_action, {action :: kz_term:api_binary()
```
Comment (jamesaimonetti) [inline applications/callflow/src/callflow.hrl]: I don't think `bl` is sufficient; I would prefer to see `blacklist` to immediately remind the reader what is being written about here.

File: applications/callflow/src/cf_blacklist.erl [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+-export([lookup/4]).
+
+-include("callflow.hrl").
+
+-record(pattern, {action :: kz_term:ne_binary()
+                 ,caller_name :: kz_term:api_ne_binary()
+                 ,has_groups :: boolean()
+                 ,names = [] :: kz_term:ne_binaries()
+                 ,regex :: re:mp()
```
Comment (jamesaimonetti) [inline applications/callflow/src/cf_blacklist.erl]: `re:mp() | 'undefined'` I think; 'undefined' must be explicitly added now

File: applications/callflow/src/cf_blacklist.erl [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+%%%-----------------------------------------------------------------------------
+-module(cf_blacklist).
+
+-export([lookup/4]).
+
+-include("callflow.hrl").
+
+-record(pattern, {action :: kz_term:ne_binary()
+                 ,caller_name :: kz_term:api_ne_binary()
+                 ,has_groups :: boolean()
```
Comment (jamesaimonetti) [inline applications/callflow/src/cf_blacklist.erl]: Should either default this to true or false or update type to `boolean() | 'undefined'`.

File: applications/callflow/src/cf_exe.erl [hunk 4]
```diff
@@ -805,7 +805,52 @@ update_actions(Action, Call) ->
...
+            cf_module_task(CFModule, Data, UpdatedCall, AMQPWorker);
+        #cf_bl_action{caller_name = CallerName, action = ActionValue} ->
+            UpdatedCall1 = kapps_call:set_user_bl_action(ActionValue, Call),
+            UpdatedCall2 = kapps_call:set_caller_id_name(CallerName, UpdatedCall1),
+            cf_module_task(CFModule, Data, UpdatedCall2, AMQPWorker)
+    end.
+
+-spec cf_try_user_blacklists(atom(), kz_json:object(), kz_term:api_ne_binary(), kapps_call:call()) -> cf_bl_action().
+cf_try_user_blacklists(CFModule, Data, AccountBlAction, Call) ->
+    try CFModule:bl_action(Data, Call) of
```
Comment (jamesaimonetti) [inline applications/callflow/src/cf_exe.erl]: If this is going to be applied to all callflow action modules, it should be an optional behaviour callback of `gen_cf_action` and the code here should check if the function is exported (using `kz_module:is_exported/3`) before calling it. The only define the function in the callflow actions that actually support it.

File: applications/callflow/src/cf_blacklist.erl [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        _ -> {'ok', Action, CallerName, Number =:= ?NO_MATCH_BL}
+    end.
+
+-spec maybe_use_nomatch(kz_term:ne_binary(), kz_term:ne_binary(), kz_term:api_ne_binary(), kz_term:ne_binary()) -> lookup_ret_number().
+maybe_use_nomatch(Number, AccountId, OwnerId, Strategy) ->
+    case knm_converters:is_reconcilable(Number, AccountId) of
+        'true' -> try_nomatch(Number, AccountId, OwnerId, Strategy);
+        'false' ->
+            lager:info("can't use no_match: number not all digits: ~s", [Number]),
+            cache_number(AccountId, OwnerId, Number, 'undefined', 'undefined')
```
Comment (jamesaimonetti) [inline applications/callflow/src/cf_blacklist.erl]: This breaks the contract with `cache_number/5`. Looks like `Action` is expected to be `ne_binary()` in the spec.

File: applications/callflow/src/cf_blacklist.erl [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    end.
+
+-spec try_nomatch(kz_term:ne_binary(), kz_term:ne_binary(), kz_term:api_ne_binary(), kz_term:ne_binary()) -> lookup_ret_number().
+try_nomatch(Number, AccountId, OwnerId, Strategy) ->
+    case lookup(?NO_MATCH_BL, AccountId, OwnerId, Strategy) of
+        {'error', _} -> cache_number(AccountId, OwnerId, Number, 'undefined', 'undefined');
```
Comment (jamesaimonetti) [inline applications/callflow/src/cf_blacklist.erl]: Similar contract violation here.

File: core/kazoo_documents/src/kzd_blacklists.erl [hunk 7]
```diff
@@ -5,23 +5,271 @@
...
+      ,{'key', ['null', Type, Number]}
+      ]);
+get_view_options(?LIST_BY_OWNER, OwnerId, <<"number">> = Type, Number) ->
+    props:filter_undefined(
+      [{'group', 'true'}
+      ,{'group_level', 3}
+      ,{'key', [OwnerId, Type, Number]}
+      ]).
+
+-spec format_view_results(kz_json:objects(), map()) -> kz_json:object().
```
Comment (jamesaimonetti) [inline core/kazoo_documents/src/kzd_blacklists.erl]: return type is wrong, this does not return `kz_json:object()` but `{'ok', kz_json:object()} | {'error', 'not_found'}`

File: core/kazoo_documents/src/kzd_blacklists.erl [hunk 8]
```diff
@@ -5,23 +5,271 @@
...
+format_view_results(JObjs, Options) ->
+    format_view_results(JObjs, kz_json:new(), Options).
+
+-spec format_view_results(kz_json:objects(), kz_json:object(), map()) -> kz_json:object().
```
Comment (jamesaimonetti) [inline core/kazoo_documents/src/kzd_blacklists.erl]: Same bad return type

File: core/kazoo_documents/src/kzd_blacklists.erl [hunk 9]
```diff
@@ -46,6 +294,99 @@ numbers(Doc, Default) ->
...
+
+-spec owner_id(doc()) -> kz_term:api_ne_binary().
+owner_id(Doc) ->
+    owner_id(Doc, 'undefined').
+
+-spec owner_id(doc(), Default) -> kz_term:ne_binary() | Default.
+owner_id(Doc, Default) ->
+    kz_json:get_ne_binary_value([<<"owner_id">>], Doc, Default).
+
+-spec set_owner_id(doc(), kz_term:ne_binary()) -> doc().
```
Comment (jamesaimonetti) [inline core/kazoo_documents/src/kzd_blacklists.erl]: Contract disagrees with the implementation; looks like `api_ne_binary()` is allowed.

File: core/kazoo_documents/src/kzd_blacklists.erl [hunk 10]
```diff
@@ -46,6 +294,99 @@ numbers(Doc, Default) ->
...
+set_owner_id(Doc, 'undefined') ->
+    kz_json:delete_key([<<"owner_id">>], Doc);
+set_owner_id(Doc, OwnerId) ->
+    kz_json:set_value([<<"owner_id">>], OwnerId, Doc).
+
+-spec set_owner_id(doc(), kz_term:ne_binary(), kz_term:ne_binary()) -> doc().
```
Comment (jamesaimonetti) [inline core/kazoo_documents/src/kzd_blacklists.erl]: Similarly, `api_ne_binary()` for second arg.

File: applications/callflow/src/cf_blacklist.erl (hunk 3 above)
Comment (sergey-safarov) [reply applications/callflow/src/cf_blacklist.erl]: I renamed record name. But modules functions still have name `bl_action`.

File: applications/callflow/src/callflow.hrl (hunk 1 above)
Comment (sergey-safarov) [reply applications/callflow/src/callflow.hrl]: I renamed record name. But modules functions still have name `bl_action`.

**Comment by sergey-safarov** [comment]
PR is updated.
Looks as failed tests not related to this PR.

**Comment by sergey-safarov** [comment]
PR is rebased to clean failed tests.

**Comment by lazedo** [comment]
@sergey-safarov can you explain the purpose of this PR,  and why it is touching all these files ?

**Comment by sergey-safarov** [comment]
Current blacklist implementation allow only block call. And blacklist granularity is account wide.
PR add features:
1. allow define user wide blacklist;
2. allow define action `pass` and now blacklist maybe used as white list (block all except allowed numbers);
3. allow define action `skip_human` and `cf_exe` will call only module that not dials end user phones;
4. allow define action 'ask_human' and when channel is established to user phone humans can make decision to accept or decline call.

About amount changed files.
Feature implemented in `callflow` application. This applications contains about 70 modules. To get user defined blacklist need get `user_id` for callflow data for each module. That is means need extend modules to extract user_id from callflow data.

**Comment by sergey-safarov** [comment]
Could you look first 17 core commits, one by one. Some of this related to `make apis`.
Rest commits related to crossbar, documentation and change request from 2600hz team.
[Commit is correct order](https://github.com/sergey-safarov/kazoo/commits/user_blacklist3).

File: applications/callflow/src/callflow.hrl (hunk 1 above)
Comment (jamesaimonetti) [reply applications/callflow/src/callflow.hrl]: Please update any `bl` to `blacklist`; no need for terseness here.

**Comment by jamesaimonetti** [comment]
I still want the blacklist callback tagged as optional in the behaviour and only implemented where it is actually used. That will cut down on the number of changed modules.

File: applications/callflow/src/callflow.hrl (hunk 1 above)
Comment (sergey-safarov) [reply applications/callflow/src/callflow.hrl]: Renamed except `kzd_users:bl_strategy/2`. Reason already exist `kzd_users:blacklist_strategy/2` generated by `make apis` and imported from template.
Optionally i can not import autoenerated `kzd_users:blacklist_strategy/2`.

**Comment by sergey-safarov** [comment]
@jamesaimonetti blacklist callback now optional.

**Comment by sergey-safarov** [comment]
Thanks Luis @lazedo for review.
```
i feel the implementation should be done higher at routing level and not action level
```
Think that is not correct.
Imagine that we have callflow where exist `cf_user for user 1` -> `cf_user for user 2`.
User 1 defined blacklist with action "block".
User 2 defined blacklist with action "ask_human".

Now we must
1. evaluate blacklist action for whole callflow and get result `ask_human`.
2. at call flow we must not call user 1 and must evaluate blacklist actions during execution again.

Think introducing blacklist at call routing level add complexity, but does not add useful things.

**Comment by lazedo** [comment]
@sergey-safarov that's why we need further discussion here.
kz_endpoint:merge would make it much easier

**Comment by sergey-safarov** [comment]
Yes, you a right.
I made several iteration before I able to publish PR.
Sure for 2600Hz team also need time to thing about it.

**Comment by sergey-safarov** [comment]
```
allow define action skip_human and cf_exe will call only modules that not dials end user phones;
can you provide a use case for this ?
```
I say it `soft blacklist`.
Caller is able work with [IVR](https://en.wikipedia.org/wiki/Interactive_voice_response) or voicemail but newer reach humans.

**Comment by sergey-safarov** [comment]
As option.
We can defined `callflow` module config option `apply_blacklist_at_routing`. If option equal `true` value and account blacklist action is `block`, then CDR for call is not created.

**Comment by bazemoreb** [comment]
Hey @lazedo - I'd talked with Darren and James about this whole thing, and I've attached our rough spec that we put together here.  I think it should help to kind of describe the overall process we were going for and also contains some of the back and forth with Darren from the Jira ticket I'd done.   Let me know!  
[User Level Screening.docx](https://github.com/2600hz/kazoo/files/3370660/User.Level.Screening.docx)

**Comment by jamesaimonetti** [comment]
rebase and correct conflicts?

**Comment by lazedo** [comment]
@sergey-safarov @bazemoreb lets try to resume this after the repo split

**Pull request closed without merging**